### ISSUE TITLE
Decorator becomes untyped when it accepts an object of the same type as a certain global.

### ISSUE THREAD

**Issue by jonathanslenders** [python/mypy#14005]
**Bug Report**

A decorator becomes untyped when it accepts an argument for which we have another instance, globally defined and referred to in the same scope.

The comments in the example should make the bug much more clear.
I had no idea how to describe this issue or search for it, so no idea whether it was reported before.

**To Reproduce**

```python
from typing import Callable


class Something:
    pass


def decorator(buffer: Something) -> Callable[[Callable[[], None]], Callable[[], None]]:
    return lambda x: x


def main() -> None:
    something  # Comment out this line and everything is fine.
    another_something = Something()

    @decorator(another_something)
    def func() -> None:
        return


something = Something()  # Move this above `main` and everything is fine.
```

**Expected Behavior**

The code snippet should pass.

**Actual Behavior**

```
 $ mypy --strict /tmp/test.py
test.py:16: error: Untyped decorator makes function "func" untyped
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 0.982
- Mypy command-line flags: --strict
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.9.13
